handlers/secrets.go

```diff
@@ -28,30 +29,70 @@ const (
 
 // MakeSecretHandler makes a handler for Create/List/Delete/Update of
 //secrets in the Kubernetes API
-func MakeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+func MakeSecretHandler(defaultNamespace string, kube kubernetes.Interface) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if r.Body != nil {
 			defer r.Body.Close()
 		}
 
-		client := kube.CoreV1().Secrets(namespace)
+		lookupNamespace, err, errCode := getLookupNamespace(defaultNamespace, r)
+		if err != nil {
+			http.Error(w, err.Error(), errCode)
+		}
+
+		client := kube.CoreV1().Secrets(lookupNamespace)
 
 		switch r.Method {
 		case http.MethodGet:
 			getSecretsHandler(client, w, r)
 		case http.MethodPost:
-			createSecretHandler(client, namespace, w, r)
+			createSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodPut:
-			replaceSecretHandler(client, namespace, w, r)
+			replaceSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodDelete:
-			deleteSecretHandler(client, namespace, w, r)
+			deleteSecretHandler(client, lookupNamespace, w, r)
 		default:
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 	}
 }
 
+// getLookupNamespace returns the namespace specified in any request type
+func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error, int) {
```

Comment:
Another modification that would generalize this is to convert it to an interface and put the `defaultNamespace` and excluded namespaces into a struct.  This would look like 

```go
type NamespaceResolver interface {
    Lookup(r *http.Request) (string, error, int)
}

type namespaceResolver struct {
    defaultNamespace string
    exclude map[string]bool
}

func NewNamespaceResolver(defaultNamespace string, exclude ...string) NamespaceResolver {
    r = namespaceResolver{
        defaultNamespace: defaultNamespace
        exclude: map[string]bool{"kube-system": true}
    }

    for _, n := range exclude {
        r.exclude[n] = true
    }

    return &r
}

func (r *namespaceResolver) Lookup(r *http.Request) (string, error, int) {
    //
    return "", nil, 0
}
```

This allows use to extend the excluded list, for example via a flag during startup.  Also, using a `map[string]bool` allows you to keep the exclusion check very simple

```go
if _, ok := r.exclude[lookupNamespace]; ok {
    return "", fmt.Errorf("unable to manage secrets within the %s namespace", lookupNamespace), http.StatusUnauthorized
}
```

For now the the namespace resolver would simply be initialized  without any extra exclude values:

```go
namespaceResolver := NewNamespaceResolver(defaultNamespace)
```

Which could happen either at the start of the `MakeSecretHandler` or could be initialized in the server setup and then passed in.


_If_ we opted to pass the resolver to the Handler, then it would look roughly like this

```go
func MakeSecretHandler(namespaceResolver NamespaceResolver, kube kubernetes.Interface) http.HandlerFunc { }
```
Passing it in will ensure that it is easy to mock in tests and the behavior can be defined and controlled globally for the server.

There is a lot to digest here, i am not saying we must do it this way.  Just a suggestion to consider.

Reply:
Although I like the solution, I think it becomes too complex for now, since we are just ignoring one specific namespace. IMHO using a simple `getLookupNamespace` makes it easier to reason about the code.

Shall we go the way I coded it?

Reply:
That is fine, but I was thinking more on this as well and there are two other aspects that we should address before we merge.  I will add those as separate comments.